{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# XPath e XQuery"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Base de dados e proposta\n",
    "A base escolhida para essa análise é a : http://tbdb.bu.edu/tbdb_sysbio/Downloads.html\n",
    "\n",
    "A TB database foi criada para auxiliar na pesquisa sobre a tuberculose. Ela contém informações referentes a diversos cromossomos relacionados à doença. Para esse trabalho, foram selecionados os dados de apenas um cromossomo (visto que a base era muito grande para trabalhar com ela inteira), e acreditamos que algumas pesquisas úteis poderiam ser realizadas com essas informações.\n",
    "\n",
    "O primeiro passo para trabalhar com essa base foi arrumar os dados da base, que foram baixados no formato GTF (.gtf), para um arquivo XML (.xml). O Gene Transfer Format (GTF) é um formato de arquivo feito especialmente para guardar dados referentes a estrutura de genes. Para essa conversão, foi utilizado o seguinte código em C++ :\n",
    "\n",
    "OBS: abrir a célula de markdown para conseguir ler o código.\n",
    "\n",
    "#include <iostream>\n",
    "#include <fstream>\n",
    "#include <string>\n",
    "#include <stdio.h>\n",
    "#include<string.h>\n",
    "\n",
    "using namespace std;\n",
    "\n",
    "//componente para o vertex\n",
    "struct GFT_LINE{\n",
    "  string seqname;\n",
    "  string type;\n",
    "  string source;\n",
    "  string feature;\n",
    "  string start_value;\n",
    "  string end_value;\n",
    "  string score ;\n",
    "  string strand;\n",
    "  string frame;\n",
    "  string gene_id;\n",
    "  string transcript_id;\n",
    "};\n",
    "\n",
    "int main()\n",
    "{\n",
    "  string line;\n",
    "  ifstream myfile (\"rhodobacter_sphaeroides_2.4.1_1_transcripts.txt\");\n",
    "  ofstream outfile;\n",
    "  outfile.open (\"file.xml\");\n",
    "  outfile<<\"<?xml version=\\\"1.0\\\"?>\"<<endl;\n",
    "  if (myfile.is_open()){\n",
    "    GFT_LINE *beforeLine = new GFT_LINE;\n",
    "    beforeLine->seqname = \"\";\n",
    "    beforeLine->type = \"\";\n",
    "    GFT_LINE *l;\n",
    "    while ( getline (myfile,line) ){\n",
    "      l = new GFT_LINE;\n",
    "      string word;\n",
    "      int start = 0;\n",
    "      int part = 0;\n",
    "      for(int i = 0; i<line.length(); i++){\n",
    "        word = line.substr(start,i-start);\n",
    "        if(word == \"Chromosome_\"){\n",
    "          word = line.substr(start,i+1);\n",
    "          l->seqname = word;\n",
    "          l->type = line[i+2];\n",
    "          i +=4;\n",
    "          start = i;\n",
    "        }else if(word == \"gene_id \"){\n",
    "          start = i+1;\n",
    "          for(int j= i+2; j < line.length(); j++){\n",
    "            word = line.substr(start,j-start);\n",
    "            if(line[j]=='\\\"'){\n",
    "              i= j+2;\n",
    "              start = j+2;\n",
    "              break;\n",
    "            }\n",
    "          }\n",
    "          i ++;\n",
    "          start = i;\n",
    "          l->gene_id = word;\n",
    "        }else if(word == \"transcript_id \"){\n",
    "          start = i+1;\n",
    "          for(int j= i+2; j < line.length(); j++){\n",
    "            word = line.substr(start,j-start);\n",
    "            if(line[j]=='\\\"'){\n",
    "              break;\n",
    "            }\n",
    "          }\n",
    "          l->transcript_id = word;\n",
    "          break;\n",
    "        }else if(line[i] == '\\t'){\n",
    "           word = line.substr(start,i-start);\n",
    "           start = i+1;\n",
    "            if(part == 0){\n",
    "              l->source = word;\n",
    "            }else if(part == 1){\n",
    "              l->feature = word;\n",
    "            }else if(part == 2){\n",
    "              l->start_value = word;\n",
    "            }else if (part == 3){\n",
    "              l->end_value = word;\n",
    "            }else if (part == 4){\n",
    "              l->score = word;\n",
    "            }else if (part == 5){\n",
    "              l->strand = word;\n",
    "            }else if (part == 6){\n",
    "              l->frame = word;\n",
    "            }\n",
    "            part++;\n",
    "        }\n",
    "      }\n",
    "      int novo = 0;\n",
    "      if(beforeLine->seqname != l->seqname){\n",
    "        if(beforeLine->seqname != \"\"){\n",
    "          outfile << \"\\t</type >\" <<endl;\n",
    "          outfile << \"</\" << beforeLine->seqname << \">\" << endl;\n",
    "        }\n",
    "        outfile << \"<\" << l->seqname << \">\" << endl;\n",
    "      }\n",
    "      if(beforeLine->type != l->type || beforeLine->seqname != l->seqname){\n",
    "        if(beforeLine->seqname != \"\"){\n",
    "          outfile << \"\\t</type >\" <<endl;\n",
    "        }\n",
    "        outfile << \"\\t<type value=\\\"\"<< l->type << \"\\\">\" <<endl;\n",
    "        novo = 1;\n",
    "      }else\n",
    "        novo = 0;\n",
    "      outfile<<\"\\t\\t<property \" ;\n",
    "      outfile<<\"source=\\\"\"<< l->source<< \"\\\" \";\n",
    "      outfile<<\"feature=\\\"\"<< l->feature<< \"\\\" \";\n",
    "      outfile<<\"start=\\\"\"<< l->start_value<< \"\\\" \";\n",
    "      outfile<<\"end=\\\"\"<< l->end_value<< \"\\\" \";\n",
    "      outfile<<\"score=\\\"\"<< l->score<< \"\\\" \";\n",
    "      outfile<<\"strand=\\\"\"<< l->strand<< \"\\\" \";\n",
    "      outfile<<\"frame=\\\"\"<< l->frame<< \"\\\" \";\n",
    "      outfile<<\"gene_id=\\\"\"<< l->gene_id<< \"\\\" \";\n",
    "      outfile<<\"transcript_id=\\\"\"<< l->transcript_id<< \"\\\" />\" << endl;\n",
    "      if((beforeLine->type != l->type && beforeLine->seqname == l->seqname) && !novo)\n",
    "        outfile << \"\\t</type >\" <<endl;\n",
    "      beforeLine = l;\n",
    "    }\n",
    "    outfile << \"\\t</type >\" <<endl;\n",
    "    outfile << \"</\" << l->seqname << \">\" << endl;\n",
    "    myfile.close();\n",
    "  }\n",
    "\n",
    "  else cout << \"Unable to open file\";\n",
    "\n",
    "    return 0;\n",
    "}\n",
    "\n",
    "\n",
    "Vale ressaltar que o código exige que o arquivo a ser convertido esteja no mesmo diretório que a main, e que sua extensão seja mudada para .txt antes da conversão.\n",
    "\n",
    "O arquivo XML criado contém os seguintes dados:\n",
    "    * source\n",
    "    * feature\n",
    "    * start\n",
    "    * end\n",
    "    * score\n",
    "    * frame\n",
    "    * attribute"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Proposta 1:\n",
    "Uma primeira informação que pode ser interessante na pesquisa de um cromossomo é sua complexidade. Uma possível forma de se medir isso utilizando a base em XML seria contar o número de propriedades que um determinado tipo de um cromossomo possui.\n",
    "\n",
    "```xquery\n",
    "let $doc := doc(\"file2.xml\")\n",
    "\n",
    "return count($doc//type[@value = 1]/property)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Resposta:** 3340"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Proposta 2:\n",
    "Contar o número de códons de parada em um determinado tipo do cromossomo.\n",
    "\n",
    "```xquery\n",
    "let $doc := doc(\"file2.xml\")\n",
    "\n",
    "for $c in ($doc//type[@value = 2])\n",
    "\n",
    "return count($c/property[@feature = 'stop_codon'])\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Resposta:** 79"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Proposta 3:\n",
    "Encontrar propriedades a determinado gene em um dado tipo de cromossomo.\n",
    "\n",
    "```xquery\n",
    "let $doc := doc(\"file2.xml\")\n",
    "\n",
    "for $c in ($doc//type[@value = 4]/property)\n",
    "where $c/@gene_id = 'RSP_2483'\n",
    "\n",
    "return $c\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Algumas Respostas:** \n",
    "```xml\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"start_codon\" start=\"71408\" end=\"71410\" score=\".\" strand=\"+\" frame=\"0\" gene_id=\"RSP_2483\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"stop_codon\" start=\"72713\" end=\"72715\" score=\".\" strand=\"+\" frame=\"0\" gene_id=\"RSP_2483\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"exon\" start=\"71408\" end=\"72715\" score=\".\" strand=\"+\" frame=\".\" gene_id=\"RSP_2483\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"CDS\" start=\"71408\" end=\"72712\" score=\".\" strand=\"+\" frame=\"0\" gene_id=\"RSP_2483\" transcript_id=\"null\"/>\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Proposta 4:\n",
    "Encontrar propriedades de um tipo de cromossomo que comecem após uma determinada posição do cromossomo.\n",
    "\n",
    "```xquery\n",
    "let $doc := doc(\"file2.xml\")\n",
    "\n",
    "for $c in ($doc//type[@value = 5]/property)\n",
    "where $c/@start > 12000\n",
    "\n",
    "return $c\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Algumas Respostas:** \n",
    "```xml\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"stop_codon\" start=\"12384\" end=\"12386\" score=\".\" strand=\"+\" frame=\"0\" gene_id=\"RSP_4150\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"start_codon\" start=\"12450\" end=\"12452\" score=\".\" strand=\"+\" frame=\"0\" gene_id=\"RSP_4151\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"stop_codon\" start=\"12768\" end=\"12770\" score=\".\" strand=\"+\" frame=\"0\" gene_id=\"RSP_4151\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"exon\" start=\"12450\" end=\"12770\" score=\".\" strand=\"+\" frame=\".\" gene_id=\"RSP_4151\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"CDS\" start=\"39087\" end=\"39968\" score=\".\" strand=\"-\" frame=\"0\" gene_id=\"RSP_4189\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"start_codon\" start=\"40681\" end=\"40683\" score=\".\" strand=\"+\" frame=\"0\" gene_id=\"RSP_4190\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"stop_codon\" start=\"41272\" end=\"41274\" score=\".\" strand=\"+\" frame=\"0\" gene_id=\"RSP_4190\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"exon\" start=\"40681\" end=\"41274\" score=\".\" strand=\"+\" frame=\".\" gene_id=\"RSP_4190\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"CDS\" start=\"40681\" end=\"41271\" score=\".\" strand=\"+\" frame=\"0\" gene_id=\"RSP_4190\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"start_codon\" start=\"41195\" end=\"41197\" score=\".\" strand=\"+\" frame=\"0\" gene_id=\"RSP_4191\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"stop_codon\" start=\"44135\" end=\"44137\" score=\".\" strand=\"+\" frame=\"0\" gene_id=\"RSP_4191\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"exon\" start=\"41195\" end=\"44137\" score=\".\" strand=\"+\" frame=\".\" gene_id=\"RSP_4191\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"CDS\" start=\"41195\" end=\"44134\" score=\".\" strand=\"+\" frame=\"0\" gene_id=\"RSP_4191\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"start_codon\" start=\"44599\" end=\"44601\" score=\".\" strand=\"+\" frame=\"0\" gene_id=\"RSP_4193\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"stop_codon\" start=\"45151\" end=\"45153\" score=\".\" strand=\"+\" frame=\"0\" gene_id=\"RSP_4193\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"exon\" start=\"44599\" end=\"45153\" score=\".\" strand=\"+\" frame=\".\" gene_id=\"RSP_4193\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"CDS\" start=\"44599\" end=\"45150\" score=\".\" strand=\"+\" frame=\"0\" gene_id=\"RSP_4193\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"start_codon\" start=\"45153\" end=\"45155\" score=\".\" strand=\"+\" frame=\"0\" gene_id=\"RSP_4194\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"stop_codon\" start=\"46185\" end=\"46187\" score=\".\" strand=\"+\" frame=\"0\" gene_id=\"RSP_4194\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"exon\" start=\"45153\" end=\"46187\" score=\".\" strand=\"+\" frame=\".\" gene_id=\"RSP_4194\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"CDS\" start=\"45153\" end=\"46184\" score=\".\" strand=\"+\" frame=\"0\" gene_id=\"RSP_4194\" transcript_id=\"null\"/>\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Proposta 5:\n",
    "Encontrar genes que remetem tanto a codons de início quanto a codons de parada em determinado tipo de cromossomo.\n",
    "\n",
    "```xquery\n",
    "let $doc := doc(\"file2.xml\")\n",
    "\n",
    "for $c in ($doc//type[@value = 6]/property),\n",
    "    $p in ($doc//type[@value = 6]/property)\n",
    "where $c/@feature = 'start_codon' and\n",
    "      $p/@feature = 'stop_codon' and\n",
    "      $c/@gene_id = $p/@gene_id\n",
    "\n",
    "return data($c/@gene_id)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Algumas Respostas:** \n",
    "```xml\n",
    "RSP_4254\n",
    "RSP_4264\n",
    "RSP_4265\n",
    "RSP_4271\n",
    "RSP_4278\n",
    "RSP_4284\n",
    "RSP_4285\n",
    "RSP_4289\n",
    "RSP_7352\n",
    "RSP_7359\n",
    "RSP_7360\n",
    "RSP_7398\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Proposta 6:\n",
    "Encontrar propriedades de um tipo de cromossomo que ocupem um \"espaço\" pequeno, ou seja, ocupam um trecho pequeno do cromossomo.\n",
    "\n",
    "```xquery\n",
    "let $doc := doc(\"file2.xml\")\n",
    "\n",
    "for $c in ($doc//type[@value = 1]/property)\n",
    "where $c/@end - $c/@start <= 10\n",
    "\n",
    "return $c\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Algumas Respostas:** \n",
    "```xml\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"stop_codon\" start=\"926940\" end=\"926942\" score=\".\" strand=\"-\" frame=\"0\" gene_id=\"RSP_1412\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"stop_codon\" start=\"929652\" end=\"929654\" score=\".\" strand=\"-\" frame=\"0\" gene_id=\"RSP_1413\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"start_codon\" start=\"930774\" end=\"930776\" score=\".\" strand=\"+\" frame=\"0\" gene_id=\"RSP_1414\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"stop_codon\" start=\"932860\" end=\"932862\" score=\".\" strand=\"+\" frame=\"0\" gene_id=\"RSP_1415\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"start_codon\" start=\"932884\" end=\"932886\" score=\".\" strand=\"+\" frame=\"0\" gene_id=\"RSP_1416\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"start_codon\" start=\"933170\" end=\"933172\" score=\".\" strand=\"+\" frame=\"0\" gene_id=\"RSP_1417\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"stop_codon\" start=\"934061\" end=\"934063\" score=\".\" strand=\"+\" frame=\"0\" gene_id=\"RSP_1417\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"stop_codon\" start=\"934330\" end=\"934332\" score=\".\" strand=\"-\" frame=\"0\" gene_id=\"RSP_1418\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"start_codon\" start=\"936337\" end=\"936339\" score=\".\" strand=\"-\" frame=\"0\" gene_id=\"RSP_1419\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"start_codon\" start=\"937278\" end=\"937280\" score=\".\" strand=\"-\" frame=\"0\" gene_id=\"RSP_1420\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"stop_codon\" start=\"936336\" end=\"936338\" score=\".\" strand=\"-\" frame=\"0\" gene_id=\"RSP_1420\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"stop_codon\" start=\"937519\" end=\"937521\" score=\".\" strand=\"-\" frame=\"0\" gene_id=\"RSP_1421\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"start_codon\" start=\"939226\" end=\"939228\" score=\".\" strand=\"-\" frame=\"0\" gene_id=\"RSP_1422\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"stop_codon\" start=\"939206\" end=\"939208\" score=\".\" strand=\"-\" frame=\"0\" gene_id=\"RSP_1423\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"start_codon\" start=\"941208\" end=\"941210\" score=\".\" strand=\"+\" frame=\"0\" gene_id=\"RSP_1425\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"start_codon\" start=\"39397\" end=\"39399\" score=\".\" strand=\"+\" frame=\"0\" gene_id=\"RSP_3003\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"start_codon\" start=\"39958\" end=\"39960\" score=\".\" strand=\"+\" frame=\"0\" gene_id=\"RSP_3004\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"stop_codon\" start=\"40564\" end=\"40566\" score=\".\" strand=\"+\" frame=\"0\" gene_id=\"RSP_3004\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"start_codon\" start=\"41165\" end=\"41167\" score=\".\" strand=\"-\" frame=\"0\" gene_id=\"RSP_3005\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"start_codon\" start=\"41397\" end=\"41399\" score=\".\" strand=\"+\" frame=\"0\" gene_id=\"RSP_3006\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"stop_codon\" start=\"42204\" end=\"42206\" score=\".\" strand=\"-\" frame=\"0\" gene_id=\"RSP_3007\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"start_codon\" start=\"45328\" end=\"45330\" score=\".\" strand=\"-\" frame=\"0\" gene_id=\"RSP_3012\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"stop_codon\" start=\"44920\" end=\"44922\" score=\".\" strand=\"-\" frame=\"0\" gene_id=\"RSP_3012\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"start_codon\" start=\"47046\" end=\"47048\" score=\".\" strand=\"-\" frame=\"0\" gene_id=\"RSP_3014\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"stop_codon\" start=\"47048\" end=\"47050\" score=\".\" strand=\"-\" frame=\"0\" gene_id=\"RSP_3015\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"start_codon\" start=\"48887\" end=\"48889\" score=\".\" strand=\"-\" frame=\"0\" gene_id=\"RSP_3016\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"start_codon\" start=\"49442\" end=\"49444\" score=\".\" strand=\"-\" frame=\"0\" gene_id=\"RSP_3017\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"stop_codon\" start=\"49518\" end=\"49520\" score=\".\" strand=\"-\" frame=\"0\" gene_id=\"RSP_3018\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"start_codon\" start=\"51386\" end=\"51388\" score=\".\" strand=\"-\" frame=\"0\" gene_id=\"RSP_3019\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"start_codon\" start=\"51853\" end=\"51855\" score=\".\" strand=\"-\" frame=\"0\" gene_id=\"RSP_3020\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"start_codon\" start=\"52764\" end=\"52766\" score=\".\" strand=\"-\" frame=\"0\" gene_id=\"RSP_3021\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"stop_codon\" start=\"51852\" end=\"51854\" score=\".\" strand=\"-\" frame=\"0\" gene_id=\"RSP_3021\" transcript_id=\"null\"/>\n",
    "<property source=\"RSP1_INSERTASSEMBLYFROMGENBANK_1\" feature=\"start_codon\" start=\"53250\" end=\"53252\" score=\".\" strand=\"+\" frame=\"0\" gene_id=\"RSP_3022\" transcript_id=\"null\"/>\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SQL",
   "language": "SQL",
   "name": "sql"
  },
  "language_info": {
   "codemirror_mode": "sql",
   "file_extension": ".sql",
   "mimetype": "",
   "name": "SQL",
   "nbconverter_exporter": "",
   "version": ""
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
