{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Modelos conceituais\n",
    "\n",
    "### UniProt = Human Diseases\n",
    "\n",
    "![](000-mc-er-uniprothd.png)\n",
    "\n",
    "### Breast Cancer Winconsin (Diagnosis) Data Set\n",
    "\n",
    "![](001-mc-er-wisconsin.png)\n",
    "\n",
    "### TB Database\n",
    "\n",
    "![](003-mc-er-tbdb.png)\n",
    "\n",
    "### Virus-Host DB\n",
    "\n",
    "![](004-mc-er-virushost.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Modelos lógicos\n",
    "\n",
    "Nota: apenas das bases que deram origem a consultas relacionais\n",
    "\n",
    "### UniProt = Human Diseases\n",
    "\n",
    "> **Diseases**(<u>Id</u>, Name, Mnemonic, Description)\n",
    ">\n",
    "> **DiseasesAnalysis**(Category, Amount)\n",
    ">\n",
    "> **DiseasesAnalysisAux**(Category, Amount)\n",
    " \n",
    "\n",
    "### Breast Cancer Winconsin (Diagnosis) Data Set\n",
    "\n",
    "> **Paciente**(<u>Id</u>, Diagnostico, Raio, Textura, Perimetro, Area, Suavidade, Compacidade, Concavidade, PontosConcavos, Simetria, DimensaoFractal)\n",
    ">\n",
    "> **Erro**(<u>IdPaciente</u>, Raio_erro, Textura_erro, Perimetro_erro, Area_erro, Suavidade_erro, Compacidade_erro, Concavidade_erro, PontosConcavos_erro, Simetria_erro, DimensaoFractal_erro)<br>\n",
    "> -> CHE IdPaciente -> Paciente(Id)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Código de conversão\n",
    "\n",
    "### TB Database\n",
    "\n",
    "Conversão do formato Gene Transfer Format (GTF) para XML:\n",
    "\n",
    "```cpp\n",
    "#include <iostream>\n",
    "#include <fstream>\n",
    "#include <string>\n",
    "#include <stdio.h>\n",
    "#include<string.h>\n",
    "\n",
    "using namespace std;\n",
    "\n",
    "//componente para o vertex\n",
    "struct GFT_LINE{\n",
    "  string seqname;\n",
    "  string type;\n",
    "  string source;\n",
    "  string feature;\n",
    "  string start_value;\n",
    "  string end_value;\n",
    "  string score ;\n",
    "  string strand;\n",
    "  string frame;\n",
    "  string gene_id;\n",
    "  string transcript_id;\n",
    "};\n",
    "\n",
    "int main()\n",
    "{\n",
    "  string line;\n",
    "  ifstream myfile (\"rhodobacter_sphaeroides_2.4.1_1_transcripts.txt\");\n",
    "  ofstream outfile;\n",
    "  outfile.open (\"file.xml\");\n",
    "  outfile<<\"<?xml version=\\\"1.0\\\"?>\"<<endl;\n",
    "  if (myfile.is_open()){\n",
    "    GFT_LINE *beforeLine = new GFT_LINE;\n",
    "    beforeLine->seqname = \"\";\n",
    "    beforeLine->type = \"\";\n",
    "    GFT_LINE *l;\n",
    "    while ( getline (myfile,line) ){\n",
    "      l = new GFT_LINE;\n",
    "      string word;\n",
    "      int start = 0;\n",
    "      int part = 0;\n",
    "      for(int i = 0; i<line.length(); i++){\n",
    "        word = line.substr(start,i-start);\n",
    "        if(word == \"Chromosome_\"){\n",
    "          word = line.substr(start,i+1);\n",
    "          l->seqname = word;\n",
    "          l->type = line[i+2];\n",
    "          i +=4;\n",
    "          start = i;\n",
    "        }else if(word == \"gene_id \"){\n",
    "          start = i+1;\n",
    "          for(int j= i+2; j < line.length(); j++){\n",
    "            word = line.substr(start,j-start);\n",
    "            if(line[j]=='\\\"'){\n",
    "              i= j+2;\n",
    "              start = j+2;\n",
    "              break;\n",
    "            }\n",
    "          }\n",
    "          i ++;\n",
    "          start = i;\n",
    "          l->gene_id = word;\n",
    "        }else if(word == \"transcript_id \"){\n",
    "          start = i+1;\n",
    "          for(int j= i+2; j < line.length(); j++){\n",
    "            word = line.substr(start,j-start);\n",
    "            if(line[j]=='\\\"'){\n",
    "              break;\n",
    "            }\n",
    "          }\n",
    "          l->transcript_id = word;\n",
    "          break;\n",
    "        }else if(line[i] == '\\t'){\n",
    "           word = line.substr(start,i-start);\n",
    "           start = i+1;\n",
    "            if(part == 0){\n",
    "              l->source = word;\n",
    "            }else if(part == 1){\n",
    "              l->feature = word;\n",
    "            }else if(part == 2){\n",
    "              l->start_value = word;\n",
    "            }else if (part == 3){\n",
    "              l->end_value = word;\n",
    "            }else if (part == 4){\n",
    "              l->score = word;\n",
    "            }else if (part == 5){\n",
    "              l->strand = word;\n",
    "            }else if (part == 6){\n",
    "              l->frame = word;\n",
    "            }\n",
    "            part++;\n",
    "        }\n",
    "      }\n",
    "      int novo = 0;\n",
    "      if(beforeLine->seqname != l->seqname){\n",
    "        if(beforeLine->seqname != \"\"){\n",
    "          outfile << \"\\t</type >\" <<endl;\n",
    "          outfile << \"</\" << beforeLine->seqname << \">\" << endl;\n",
    "        }\n",
    "        outfile << \"<\" << l->seqname << \">\" << endl;\n",
    "      }\n",
    "      if(beforeLine->type != l->type || beforeLine->seqname != l->seqname){\n",
    "        if(beforeLine->seqname != \"\"){\n",
    "          outfile << \"\\t</type >\" <<endl;\n",
    "        }\n",
    "        outfile << \"\\t<type value=\\\"\"<< l->type << \"\\\">\" <<endl;\n",
    "        novo = 1;\n",
    "      }else\n",
    "        novo = 0;\n",
    "      outfile<<\"\\t\\t<property \" ;\n",
    "      outfile<<\"source=\\\"\"<< l->source<< \"\\\" \";\n",
    "      outfile<<\"feature=\\\"\"<< l->feature<< \"\\\" \";\n",
    "      outfile<<\"start=\\\"\"<< l->start_value<< \"\\\" \";\n",
    "      outfile<<\"end=\\\"\"<< l->end_value<< \"\\\" \";\n",
    "      outfile<<\"score=\\\"\"<< l->score<< \"\\\" \";\n",
    "      outfile<<\"strand=\\\"\"<< l->strand<< \"\\\" \";\n",
    "      outfile<<\"frame=\\\"\"<< l->frame<< \"\\\" \";\n",
    "      outfile<<\"gene_id=\\\"\"<< l->gene_id<< \"\\\" \";\n",
    "      outfile<<\"transcript_id=\\\"\"<< l->transcript_id<< \"\\\" />\" << endl;\n",
    "      if((beforeLine->type != l->type && beforeLine->seqname == l->seqname) && !novo)\n",
    "        outfile << \"\\t</type >\" <<endl;\n",
    "      beforeLine = l;\n",
    "    }\n",
    "    outfile << \"\\t</type >\" <<endl;\n",
    "    outfile << \"</\" << l->seqname << \">\" << endl;\n",
    "    myfile.close();\n",
    "  }\n",
    "\n",
    "  else cout << \"Unable to open file\";\n",
    "\n",
    "    return 0;\n",
    "}\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SQL\n",
    "\n",
    "Os scripts de criação de tabela e normalização dos dados estão disponíveis nos notebooks em anexo. Seguem abaixo as principais consultas realizadas nas bases de dados."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### UniProt - Human Diseases\n",
    "\n",
    "#### Doenças mais comuns\n",
    "\n",
    "Esta query em SQL, retorna as informações das 5 categorias de doenças mais comuns relacionadas a algum tipo de proteina\n",
    "\n",
    "```sql\n",
    "SELECT * FROM DiseasesAnalysis\n",
    "ORDER BY amount DESC\n",
    "LIMIT 11;\n",
    "```\n",
    "\n",
    "#### Suporte\n",
    "\n",
    "Esta query em SQL, retorna o valor do suporte de cada categoria de doença, ou seja, a probabilidade de dada uma doença qualquer, ela ser dessa categoria específica\n",
    "\n",
    "```sql\n",
    "SELECT Category, SUM(CAST(Amount AS DOUBLE) / (SELECT COUNT(*) numberofdiseases\n",
    "                                               FROM Diseases)) suporte\n",
    "FROM DiseasesAnalysis\n",
    "GROUP BY category\n",
    "ORDER BY suporte DESC;\n",
    "```\n",
    "\n",
    "#### Duas categorias na mesma doença\n",
    "\n",
    "Uma informação interessante a se pensar com relação aos dados de doençar é a quantidade de doenças que pertencem a duas categorias simultaneamente. Essa query mostra essas ocorrências pra algumas das categorias mais comuns escolhidas para analise\n",
    "\n",
    "```sql\n",
    "SELECT 'syndrome', 'deficiency', COUNT(*) occurrences\n",
    "FROM Diseases D\n",
    "WHERE D.name LIKE '%syndrome%' AND\n",
    "      D.name LIKE '%deficiency%';\n",
    "      \n",
    "SELECT 'syndrome', 'mental', COUNT(*) occurrences\n",
    "FROM Diseases D\n",
    "WHERE D.name LIKE '%syndrome%' AND\n",
    "      D.name LIKE '%mental%';\n",
    "      \n",
    "SELECT 'syndrome', 'congenital', COUNT(*) occurrences\n",
    "FROM Diseases D\n",
    "WHERE D.name LIKE '%syndrome%' AND\n",
    "      D.name LIKE '%congenital%';\n",
    "      \n",
    "SELECT 'mental', 'deficiency', COUNT(*) occurrences\n",
    "FROM Diseases D\n",
    "WHERE D.name LIKE '%mental%' AND\n",
    "      D.name LIKE '%deficiency%';\n",
    "      \n",
    "SELECT 'autosomal dominant', 'deafness', COUNT(*) occurrences\n",
    "FROM Diseases D\n",
    "WHERE D.name LIKE '%autosomal dominant%' AND\n",
    "      D.name LIKE '%deafness%';\n",
    "      \n",
    "SELECT 'autosomal dominant', 'syndrome', COUNT(*) occurrences\n",
    "FROM Diseases D\n",
    "WHERE D.name LIKE '%autosomal dominant%' AND\n",
    "      D.name LIKE '%syndrome%';\n",
    "      \n",
    "SELECT 'autosomal dominant', 'mental', COUNT(*) occurrences\n",
    "FROM Diseases D\n",
    "WHERE D.name LIKE '%autosomal dominant%' AND\n",
    "      D.name LIKE '%mental%';\n",
    "      \n",
    "SELECT 'X-linked', 'dysplasia', COUNT(*) occurrences\n",
    "FROM Diseases D\n",
    "WHERE D.name LIKE '%X-linked%' AND\n",
    "      D.name LIKE '%dysplasia%';\n",
    "      \n",
    "SELECT 'hereditary', 'syndrome', COUNT(*) occurrences\n",
    "FROM Diseases D\n",
    "WHERE D.name LIKE '%hereditary%' AND\n",
    "      D.name LIKE '%syndrome%';\n",
    "```\n",
    "\n",
    "#### Confiança\n",
    "\n",
    "Outra informação possivelmente mais importante que as ocorrências de duas categorias ao mesmo tempo, é a confiança dessas associações: a probabilidade de que ambas ocorram ao mesmo tempo\n",
    "\n",
    "```sql\n",
    "SELECT 'syndrome', 'deficiency', CAST((SELECT COUNT(*) occurrences\n",
    "                                       FROM Diseases D\n",
    "                                       WHERE D.name LIKE '%syndrome%' AND\n",
    "                                             D.name LIKE '%deficiency%') AS DOUBLE) / (SELECT amount\n",
    "                                                                                       FROM DiseasesAnalysis\n",
    "                                                                                       WHERE category = 'syndrome') confiança;\n",
    "                                                             \n",
    "SELECT 'syndrome', 'mental', CAST((SELECT COUNT(*) occurrences\n",
    "                                   FROM Diseases D\n",
    "                                   WHERE D.name LIKE '%syndrome%' AND\n",
    "                                         D.name LIKE '%mental%') AS DOUBLE) / (SELECT amount\n",
    "                                                                               FROM DiseasesAnalysis\n",
    "                                                                               WHERE category = 'syndrome') confiança;\n",
    "                                                             \n",
    "SELECT 'syndrome', 'congenital', CAST((SELECT COUNT(*) occurrences\n",
    "                                       FROM Diseases D\n",
    "                                       WHERE D.name LIKE '%syndrome%' AND\n",
    "                                             D.name LIKE '%congenital%') AS DOUBLE) / (SELECT amount\n",
    "                                                                                       FROM DiseasesAnalysis\n",
    "                                                                                       WHERE category = 'syndrome') confiança;\n",
    "                                                             \n",
    "SELECT 'mental', 'deficiency', CAST((SELECT COUNT(*) occurrences\n",
    "                                     FROM Diseases D\n",
    "                                     WHERE D.name LIKE '%mental%' AND\n",
    "                                           D.name LIKE '%deficiency%') AS DOUBLE) / (SELECT amount\n",
    "                                                                                     FROM DiseasesAnalysis\n",
    "                                                                                     WHERE category = 'mental') confiança;\n",
    "                                                             \n",
    "SELECT 'autosomal dominant', 'deafness', CAST((SELECT COUNT(*) occurrences\n",
    "                                               FROM Diseases D\n",
    "                                               WHERE D.name LIKE '%autosomal dominant%' AND\n",
    "                                                     D.name LIKE '%deafness%') AS DOUBLE) / (SELECT amount\n",
    "                                                                                             FROM DiseasesAnalysis\n",
    "                                                                                             WHERE category = 'autosomal dominant') confiança;\n",
    "                                                             \n",
    "SELECT 'autosomal dominant', 'syndrome', CAST((SELECT COUNT(*) occurrences\n",
    "                                               FROM Diseases D\n",
    "                                               WHERE D.name LIKE '%autosomal dominant%' AND\n",
    "                                                     D.name LIKE '%syndrome%') AS DOUBLE) / (SELECT amount\n",
    "                                                                                             FROM DiseasesAnalysis\n",
    "                                                                                             WHERE category = 'autosomal dominant') confiança;\n",
    "                                                             \n",
    "SELECT 'autosomal dominant', 'mental', CAST((SELECT COUNT(*) occurrences\n",
    "                                             FROM Diseases D\n",
    "                                             WHERE D.name LIKE '%autosomal dominant%' AND\n",
    "                                                   D.name LIKE '%mental%') AS DOUBLE) / (SELECT amount\n",
    "                                                                                         FROM DiseasesAnalysis\n",
    "                                                                                         WHERE category = 'autosomal dominant') confiança;\n",
    "                                                             \n",
    "SELECT 'X-linked', 'dysplasia', CAST((SELECT COUNT(*) occurrences\n",
    "                                      FROM Diseases D\n",
    "                                      WHERE D.name LIKE '%X-linked%' AND\n",
    "                                            D.name LIKE '%dysplasia%') AS DOUBLE) / (SELECT amount\n",
    "                                                                                     FROM DiseasesAnalysis\n",
    "                                                                                     WHERE category = 'X-linked') confiança;\n",
    "```\n",
    "\n",
    "#### Probabilidade\n",
    "\n",
    "Outra informação que podemos retirar comparar as probabilidade de uma doença ser de uma categoria mais especifica em relação a uma menos especifica.\n",
    "\n",
    "```sql\n",
    "SELECT CAST(SELECT SUM(amount) from DiseasesAnalysis \n",
    "            WHERE category LIKE '%autosomal dominant%' AS DOUBLE)/\n",
    "            CAST((SELECT count(*) NAME \n",
    "                  FROM Diseases \n",
    "                  WHERE name LIKE '%autosomal%') AS DOUBLE) autosomalDominant_autosomal,\n",
    "        CAST(SELECT SUM(amount) from DiseasesAnalysis \n",
    "            WHERE category LIKE '%autosomal dominant%' AS DOUBLE)/\n",
    "            CAST((SELECT count(*) NAME \n",
    "                  FROM Diseases) AS DOUBLE) autosomalDominant_total;\n",
    "\n",
    "SELECT (CAST(SELECT SUM(amount) from DiseasesAnalysis \n",
    "            WHERE category LIKE '%autosomal recessive%' AS DOUBLE)/\n",
    "            CAST((SELECT count(*) NAME \n",
    "                  FROM Diseases \n",
    "                  WHERE name LIKE '%autosomal%') AS DOUBLE))autosomalRecessive_autosomal,\n",
    "       (CAST(SELECT SUM(amount) from DiseasesAnalysis \n",
    "            WHERE category LIKE '%autosomal recessive%' AS DOUBLE)/\n",
    "            CAST((SELECT count(*) NAME \n",
    "                  FROM Diseases) AS DOUBLE)) autosomalRecessive_total;\n",
    "\n",
    "SELECT (CAST(SELECT SUM(amount) from DiseasesAnalysis \n",
    "            WHERE category LIKE '%x-linked%' AS DOUBLE)/\n",
    "            CAST((SELECT count(*) NAME \n",
    "                  FROM Diseases \n",
    "                  WHERE name LIKE '%linked%') AS DOUBLE)) X_LINKED,\n",
    "       (CAST(SELECT SUM(amount) from DiseasesAnalysis \n",
    "            WHERE category LIKE '%x-linked%' AS DOUBLE)/\n",
    "            CAST((SELECT count(*) NAME \n",
    "                  FROM Diseases) AS DOUBLE)) X_TOTAL;\n",
    "                  \n",
    "SELECT (CAST(SELECT SUM(amount) from DiseasesAnalysis \n",
    "            WHERE category LIKE '%y-linked%' AS DOUBLE)/\n",
    "            CAST((SELECT count(*) NAME \n",
    "                  FROM Diseases \n",
    "                  WHERE name LIKE '%linked%') AS DOUBLE)) Y_LINKED,\n",
    "       (CAST(SELECT SUM(amount) from DiseasesAnalysis \n",
    "            WHERE category LIKE '%y-linked%' AS DOUBLE)/\n",
    "            CAST((SELECT count(*) NAME \n",
    "                  FROM Diseases) AS DOUBLE)) Y_TOTAL;\n",
    "\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Breast Cancer Winconsin (Diagnosis) Data Set\n",
    "\n",
    "Consulta simples para descobrirmos as medias de valores das amostras diagnosticadas como benigmas ou malignas, para termos ideia de alguns limiares usados para analise dos dados\n",
    "\n",
    "```sql\n",
    "SELECT AVG(Raio) AS raio_media, AVG(Textura) AS textura_media, AVG(Perimetro) AS perimetro_media, AVG(Area) AS area_media,\n",
    "       AVG(Suavidade) AS suavidade_media, AVG(Compacidade) AS compacidade_media, AVG(Concavidade) AS concavidade_media,\n",
    "       AVG(PontosConcavos) AS pontosconcavos_media, AVG(Simetria) AS simetria_media, \n",
    "       AVG(DimensaoFractal) AS dimensaofractal_media\n",
    "FROM Paciente\n",
    "WHERE Diagnostico = 'M';\n",
    "\n",
    "SELECT AVG(Raio) AS raio_media, AVG(Textura) AS textura_media, AVG(Perimetro) AS perimetro_media, AVG(Area) AS area_media,\n",
    "       AVG(Suavidade) AS suavidade_media, AVG(Compacidade) AS compacidade_media, AVG(Concavidade) AS concavidade_media,\n",
    "       AVG(PontosConcavos) AS pontosconcavos_media, AVG(Simetria) AS simetria_media, \n",
    "       AVG(DimensaoFractal) AS dimensaofractal_media\n",
    "FROM Paciente\n",
    "WHERE Diagnostico = 'B';\n",
    "```\n",
    "\n",
    "Outra consulta simples em que recuperamos os valores maximos e minimos para compararmos com as medias obtidas acima, onde talvez possamos obter alguns dados úteis\n",
    "\n",
    "```sql\n",
    "SELECT MAX(Raio) AS raio_max, MAX(Textura) AS textura_max, MAX(Perimetro) AS perimetro_max, MAX(Area) AS area_max,\n",
    "       MAX(Suavidade) AS suavidade_max, MAX(Compacidade) AS compacidade_max, MAX(Concavidade) AS concavidade_max,\n",
    "       MAX(PontosConcavos) AS pontosconcavos_max, MAX(Simetria) AS simetria_max, \n",
    "       MAX(DimensaoFractal) AS dimensaofractal_max\n",
    "FROM Paciente\n",
    "WHERE Diagnostico = 'M';\n",
    "\n",
    "SELECT MAX(Raio) AS raio_max, MAX(Textura) AS textura_max, MAX(Perimetro) AS perimetro_max, MAX(Area) AS area_max,\n",
    "       MAX(Suavidade) AS suavidade_max, MAX(Compacidade) AS compacidade_max, MAX(Concavidade) AS concavidade_max,\n",
    "       MAX(PontosConcavos) AS pontosconcavos_max, MAX(Simetria) AS simetria_max, \n",
    "       MAX(DimensaoFractal) AS dimensaofractal_max\n",
    "FROM Paciente\n",
    "WHERE Diagnostico = 'B';\n",
    "```\n",
    "\n",
    "Consultas que checas quantas amostras benigmas possuem determinada característica maior que a média de amostras malignas, para determinar se tal característica é um fator importante para indicar se é benigno ou maligno.\n",
    "\n",
    "```sql\n",
    "SELECT COUNT(*)\n",
    "FROM Paciente Pa\n",
    "WHERE Pa.Diagnostico = 'B' AND Pa.Area > (SELECT AVG(P.Area) FROM Paciente P WHERE P.Diagnostico = 'M');\n",
    "\n",
    "SELECT COUNT(*)\n",
    "FROM Paciente Pa\n",
    "WHERE Pa.Diagnostico = 'B' AND Pa.Perimetro > (SELECT AVG(P.Perimetro) FROM Paciente P WHERE P.Diagnostico = 'M');\n",
    "\n",
    "SELECT COUNT(*)\n",
    "FROM Paciente Pa\n",
    "WHERE Pa.Diagnostico = 'B' AND Pa.Concavidade > (SELECT AVG(P.Concavidade) FROM Paciente P WHERE P.Diagnostico = 'M');\n",
    "\n",
    "SELECT COUNT(*)\n",
    "FROM Paciente Pa\n",
    "WHERE Pa.Diagnostico = 'B' AND Pa.Compacidade > (SELECT AVG(P.Compacidade) FROM Paciente P WHERE P.Diagnostico = 'M');\n",
    "\n",
    "SELECT COUNT(*)\n",
    "FROM Paciente Pa\n",
    "WHERE Pa.Diagnostico = 'B' AND Pa.PontosConcavos > (SELECT AVG(P.PontosConcavos) FROM Paciente P WHERE P.Diagnostico = 'M');\n",
    "```\n",
    "\n",
    "Consulta para verificar uma amostra considerando os valores obtidos nos 5 fatores acima, de forma a identificar quantas amostras benignas fogem dos padrões identificados\n",
    "\n",
    "```sql\n",
    "SELECT COUNT(*)\n",
    "FROM Paciente\n",
    "WHERE Diagnostico = 'B' AND Area > 978.37 AND Perimetro > 115.36 AND Concavidade > 0.16 AND Compacidade >  0.145 AND \n",
    "      PontosConcavos > 0.088\n",
    "```\n",
    "\n",
    "Cálculo de correlações entre área e demais características\n",
    "\n",
    "```sql\n",
    "\n",
    "-- raio X área\n",
    "SELECT ((N*SumTot)-(SumArea*SumPer))/(SQRT(N*SumArea2-SumArea*sumArea)*SQRT(N*SumPer2-SumPer*sumPer))\n",
    "FROM   (SELECT SUM(Area) AS SumArea,\n",
    "               SUM(Raio) AS sumPer,\n",
    "               SUM(Area*Area) AS SumArea2,\n",
    "               SUM(Raio*Raio) AS sumPer2,\n",
    "               SUM(Area*Raio) AS sumTot,\n",
    "               COUNT(*) AS N\n",
    "         FROM   Paciente\n",
    "        ) dados;\n",
    "\n",
    "-- perímetro X área\n",
    "SELECT ((N*SumTot)-(SumArea*SumPer))/(SQRT(N*SumArea2-SumArea*sumArea)*SQRT(N*SumPer2-SumPer*sumPer))\n",
    "FROM   (SELECT SUM(Area) AS SumArea,\n",
    "               SUM(Perimetro) AS sumPer,\n",
    "               SUM(Area*Area) AS SumArea2,\n",
    "               SUM(Perimetro*Perimetro) AS sumPer2,\n",
    "               SUM(Area*Perimetro) AS sumTot,\n",
    "               COUNT(*) AS N\n",
    "         FROM   Paciente\n",
    "        ) dados;\n",
    "\n",
    "-- concavidade X área\n",
    "SELECT ((N*SumTot)-(SumArea*SumPer))/(SQRT(N*SumArea2-SumArea*sumArea)*SQRT(N*SumPer2-SumPer*sumPer))\n",
    "FROM   (SELECT SUM(Area) AS SumArea,\n",
    "               SUM(Concavidade) AS sumPer,\n",
    "               SUM(Area*Area) AS SumArea2,\n",
    "               SUM(Concavidade*Concavidade) AS sumPer2,\n",
    "               SUM(Area*Concavidade) AS sumTot,\n",
    "               COUNT(*) AS N\n",
    "         FROM   Paciente\n",
    "        ) dados;\n",
    "\n",
    "-- pontos concavos X área\n",
    "SELECT ((N*SumTot)-(SumArea*SumPer))/(SQRT(N*SumArea2-SumArea*sumArea)*SQRT(N*SumPer2-SumPer*sumPer))\n",
    "FROM   (SELECT SUM(Area) AS SumArea,\n",
    "               SUM(PontosConcavos) AS sumPer,\n",
    "               SUM(Area*Area) AS SumArea2,\n",
    "               SUM(PontosConcavos*PontosConcavos) AS sumPer2,\n",
    "               SUM(Area*PontosConcavos) AS sumTot,\n",
    "               COUNT(*) AS N\n",
    "         FROM   Paciente\n",
    "        ) dados;\n",
    "\n",
    "-- textura X área\n",
    "SELECT ((N*SumTot)-(SumArea*SumPer))/(SQRT(N*SumArea2-SumArea*sumArea)*SQRT(N*SumPer2-SumPer*sumPer))\n",
    "FROM   (SELECT SUM(Area) AS SumArea,\n",
    "               SUM(Textura) AS sumPer,\n",
    "               SUM(Area*Area) AS SumArea2,\n",
    "               SUM(Textura*Textura) AS sumPer2,\n",
    "               SUM(Area*Textura) AS sumTot,\n",
    "               COUNT(*) AS N\n",
    "         FROM   Paciente\n",
    "        ) dados;\n",
    "\n",
    "-- dimensão fractal X área\n",
    "SELECT ((N*SumTot)-(SumArea*SumPer))/(SQRT(N*SumArea2-SumArea*sumArea)*SQRT(N*SumPer2-SumPer*sumPer))\n",
    "FROM   (SELECT SUM(Area) AS SumArea,\n",
    "               SUM(DimensaoFractal) AS sumPer,\n",
    "               SUM(Area*Area) AS SumArea2,\n",
    "               SUM(DimensaoFractal*DimensaoFractal) AS sumPer2,\n",
    "               SUM(Area*DimensaoFractal) AS sumTot,\n",
    "               COUNT(*) AS N\n",
    "         FROM   Paciente\n",
    "        ) dados;\n",
    "\n",
    "-- suavidade X área\n",
    "SELECT ((N*SumTot)-(SumArea*SumPer))/(SQRT(N*SumArea2-SumArea*sumArea)*SQRT(N*SumPer2-SumPer*sumPer))\n",
    "FROM   (SELECT SUM(Area) AS SumArea,\n",
    "               SUM(Suavidade) AS sumPer,\n",
    "               SUM(Area*Area) AS SumArea2,\n",
    "               SUM(Suavidade*Suavidade) AS sumPer2,\n",
    "               SUM(Area*Suavidade) AS sumTot,\n",
    "               COUNT(*) AS N\n",
    "         FROM   Paciente\n",
    "        ) dados;\n",
    "\n",
    "-- simetria X área\n",
    "SELECT ((N*SumTot)-(SumArea*SumPer))/(SQRT(N*SumArea2-SumArea*sumArea)*SQRT(N*SumPer2-SumPer*sumPer))\n",
    "FROM   (SELECT SUM(Area) AS SumArea,\n",
    "               SUM(Simetria) AS sumPer,\n",
    "               SUM(Area*Area) AS SumArea2,\n",
    "               SUM(Simetria*Simetria) AS sumPer2,\n",
    "               SUM(Area*Simetria) AS sumTot,\n",
    "               COUNT(*) AS N\n",
    "         FROM   Paciente\n",
    "        ) dados;\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# XQuery"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### UniProt - UniParc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proposta 1:\n",
    "É comum que algumas pessoas tenham alergia a alguns tipos de proteina, e isso é especialmente perigoso caso se essa proteina estiver relacionada a algum tipo de organismos contidos em vacinas, portanto uma pesquisa interessante é retornar todas as vacinas relacionadas a algum tipo de proteina\n",
    "\n",
    "```xquery\n",
    "let $doc := doc(\"uniparc-250.xml\")\n",
    "for $p in $doc//*:dbReference\n",
    "   where contains(data($p/*:property[@type= 'gene_name']/@value), 'VAC') and $p/*:property[@type='NCBI_taxonomy_id']/@value != '10245'\n",
    "return $p\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proposta 2:\n",
    "Pode ser interessante analisarmos o total de proteinas relacionadas com a mesma taxonomia  \n",
    "\n",
    "```xquery\n",
    "let $doc := doc(\"uniparc-250.xml\")\n",
    "return count($doc//*:property[@type= 'NCBI_taxonomy_id'][@value= '32606'])\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proposta 3\n",
    "Agrupar todas as proteinas com um mesmo proteoma também facilita uma série de analises\n",
    "```xquery\n",
    "let $doc := doc(\"uniparc-250.xml\")\n",
    "return $doc//*:property[@type= 'proteome_id'][@value= 'UP000304398']\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proposta 4\n",
    "Encontrar todas as proteinas que tem mais de uma versão, pois isso significa que todas elas são sintéticas\n",
    "```xquery\n",
    "let $doc := doc(\"uniparc-250.xml\")\n",
    "return $doc//*:dbReference[@version_i > 1]\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proposta 5\n",
    "Encontrar todas as proteinas que tem mais de uma versão, e estão ativas\n",
    "```xquery\n",
    "let $doc := doc(\"uniparc-250.xml\")\n",
    "return $doc//*:dbReference[@version_i > 1][@active='Y']\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proposta 6\n",
    "Encontrar proteinas relacionadas a algum tipo de organismos contidos em vacinas\n",
    "```xquery\n",
    "let $doc := doc(\"uniparc-250.xml\")\n",
    "for $p in $doc//*:dbReference\n",
    "   where contains(data($p/*:property[@type= 'gene_name']/@value), 'VAC')\n",
    "return $p\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TB Database"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proposta 1:\n",
    "Uma primeira informação que pode ser interessante na pesquisa de um cromossomo é sua complexidade. Uma possível forma de se medir isso utilizando a base em XML seria contar o número de propriedades que um determinado tipo de um cromossomo possui.\n",
    "\n",
    "```xquery\n",
    "let $doc := doc(\"file2.xml\")\n",
    "\n",
    "return count($doc//type[@value = 1]/property)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proposta 2:\n",
    "Contar o número de códons de parada em um determinado tipo do cromossomo.\n",
    "\n",
    "```xquery\n",
    "let $doc := doc(\"file2.xml\")\n",
    "\n",
    "for $c in ($doc//type[@value = 2])\n",
    "\n",
    "return count($c/property[@feature = 'stop_codon'])\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proposta 3:\n",
    "Encontrar propriedades a determinado gene em um dado tipo de cromossomo.\n",
    "\n",
    "```xquery\n",
    "let $doc := doc(\"file2.xml\")\n",
    "\n",
    "for $c in ($doc//type[@value = 4]/property)\n",
    "where $c/@gene_id = 'RSP_2483'\n",
    "\n",
    "return $c\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proposta 4:\n",
    "Encontrar propriedades de um tipo de cromossomo que comecem após uma determinada posição do cromossomo.\n",
    "\n",
    "```xquery\n",
    "let $doc := doc(\"file2.xml\")\n",
    "\n",
    "for $c in ($doc//type[@value = 5]/property)\n",
    "where $c/@start > 12000\n",
    "\n",
    "return $c\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proposta 5:\n",
    "Encontrar genes que remetem tanto a codons de início quanto a codons de parada em determinado tipo de cromossomo.\n",
    "\n",
    "```xquery\n",
    "let $doc := doc(\"file2.xml\")\n",
    "\n",
    "for $c in ($doc//type[@value = 6]/property),\n",
    "    $p in ($doc//type[@value = 6]/property)\n",
    "where $c/@feature = 'start_codon' and\n",
    "      $p/@feature = 'stop_codon' and\n",
    "      $c/@gene_id = $p/@gene_id\n",
    "\n",
    "return data($c/@gene_id)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proposta 6:\n",
    "Encontrar propriedades de um tipo de cromossomo que ocupem um \"espaço\" pequeno, ou seja, ocupam um trecho pequeno do cromossomo.\n",
    "\n",
    "```xquery\n",
    "let $doc := doc(\"file2.xml\")\n",
    "\n",
    "for $c in ($doc//type[@value = 1]/property)\n",
    "where $c/@end - $c/@start <= 10\n",
    "\n",
    "return $c\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cypher"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Virus-Host DB"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Construção do grafo:\n",
    "Com as seguintes instruções abaixo realizamos a construção dos grafos, o grafo é bem simples, basicamente retiramos os virus e os hospedeiros dos mesmos, adicionamos cada um deles apenas uma única vez no banco de dados. Um certo vírus possui uma aresta para seu hospedeiro (pode ter mais de um).\n",
    "\n",
    "```cypher\n",
    "USING PERIODIC COMMIT 10\n",
    "LOAD CSV WITH HEADERS FROM\n",
    "'file:///virushostdb.csv' AS line\n",
    "WITH line\n",
    "WHERE NOT line.host_tax_id IS NULL\n",
    "MERGE (v:Virus {virus_name:line.virus_name,virus_id:line.virus_tax_id})\n",
    "MERGE (h:Host{host_id:line.host_tax_id,host_name:line.host_name})\n",
    "MERGE (v)-[i:Infects]->(h)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proposta 1:\n",
    "A primeira consulta que realizamos é determinar quantos vírus tem com seu hospedeiro o ser humano, ou seja a espécie \"Homo sapiens\".\n",
    "\n",
    "```cypher\n",
    "MATCH p=(v:Virus)-[r:Infects]->(h:Host {host_name:\"Homo sapiens\"}) \n",
    "RETURN count(v)\n",
    "```\n",
    "\n",
    "![](graph.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proposta 2:\n",
    "Aqui fazemos a análise de comunidades, apenas para percerbermos que nosso grafo não é bem adequado para esta análise,\n",
    "uma vez que o número de comunidades é o mesmo que o número de nós, logo, cada nó é uma comunidade.\n",
    "\n",
    "```cypher\n",
    "CALL algo.unionFind('Host','Infects', {\n",
    "  write: true,\n",
    "  writeProperty: 'community'\n",
    "})\n",
    "YIELD nodes, setCount, loadMillis, computeMillis, writeMillis\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SQL",
   "language": "SQL",
   "name": "sql"
  },
  "language_info": {
   "codemirror_mode": "sql",
   "file_extension": ".sql",
   "mimetype": "",
   "name": "SQL",
   "nbconverter_exporter": "",
   "version": ""
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": false,
   "sideBar": false,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": false,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
